### PR TITLE
Add systemd unit files to the RPM/DEB packages

### LINKED ISSUE #1575: Better deb packaging
### Clear and concise description of the problem

Deb packages of Woodpecker contain only a binary which basically makes the deb package near useless.

### Suggested solution

Include Systemd service file and the example config into the deb package, so Woodpecker is ready to start out of the box (or at least after quick setup by changing a couple of lines in an already included config file) on the Debian systems.

### Alternative

_No response_

### Additional context

Not everyone can use Docker in their setup. LXC containers, for example, don't support Docker.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by johanneskastl**
* add systemd unit file for server (`woodpecker-server.service`)
* add systemd unit file for agent (`woodpecker-agent.service`)
* add systemd unit files to nfpm package definitions
* add etc config file examples

fixes #1575

### CODE REVIEW

**Comment by lafriks** [comment]
For rpm/deb packages we should probably not use /usr/local but standart locations

**Comment by johanneskastl** [comment]
> For rpm/deb packages we should probably not use /usr/local but standart locations

I just extended what was already there with the systemd units.

Using `/usr/bin/` instead of `/usr/local/bin` is a different discussion. Especially as the RPM/DEB is not a full-fledged package with what most distributions expect and does not adhere to their guidelines. Basically it just packages the binary currently.

So my 2 cents would be that using `/usr/local/bin` is fine.

File: nfpm/woodpecker-agent.service [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ConditionFileNotEmpty=/etc/woodpecker/woodpecker-agent.env
+ConditionPathExists=/etc/woodpecker/woodpecker-agent.env
+
+[Service]
+Type=simple
+EnvironmentFile=/etc/woodpecker/woodpecker-agent.env
+User=woodpecker
+Group=woodpecker
+ExecStart=/usr/local/bin/woodpecker-agent
+WorkingDirectory=/var/lib/woodpecker/
```
[the remark this answers is not in the document]
Comment (johanneskastl) [reply nfpm/woodpecker-agent.service:15]: Done, see latest commits. I also rebased on latest state in main.

**Review by pat-s — COMMENTED** [review]
I don't have much experience bundling binaries + systemd but let's try to make some progress:

- I assume you created the bundles locally and tested them?
- We should extend the docs to explain that users need/can place env vars in `/etc/woodpecker/woodpecker-server.env` (and the agent counterpart).

**Comment by johanneskastl** [comment]
>     * I assume you created the bundles locally and tested them?

I tested them manually, i.e. I spun up VMs, installed the RPMs without the systemd service and then created the files locally.

The systemd units look similar to what I packaged for openSUSE in the server and agent packages, where I have a vagrant-libvirt setup for testing:

https://github.com/johanneskastl/woodpecker_vagrant_libvirt_ansible

The major difference is that I create a system user `woodpecker` for the openSUSE package and use it for the woodpecker services. This is not in this PR, as creating users from RPM/DEB is not done the same way on all distros, hence hard to package in a generic RPM/DEB. 

I did not have time to prepare a similar setup using the upstream packages and the files from this PR.
 
>     * We should extend the docs to explain that users need/can place env vars in `/etc/woodpecker/woodpecker-server.env` (and the agent counterpart).

I have no idea how creating RPM/DEB packages with nfpm works, but I think it might be a good idea to create the `/etc/woodpecker/` directory and maybe add an example file in there, to show people how the syntax in that file should be.

**Comment by zc-devs** [comment]
> creating users from RPM/DEB is not done the same way on all distros

Now, [there is a way](https://rpm-software-management.github.io/rpm/manual/users_and_groups.html).

**Comment by johanneskastl** [comment]
> > creating users from RPM/DEB is not done the same way on all distros
> 
> Now, [there is a way](https://rpm-software-management.github.io/rpm/manual/users_and_groups.html).

I know that systemd und RPM support user creation, that is how openSUSE handles this.

But the DEB and RPM packages are used on other distributions. And are built using nfpm, where I do not know how easy it is to package the files and add the necessary %pre and %post snippets.

But if someone wants to add this functionality to this PR, please feel free. I just wanted to start with a basic example that should (hopefully) work everywhere[tm].

**Comment by zc-devs** [comment]
> I know that systemd und RPM support user creation, that is how openSUSE handles this.

👍 

> But the DEB

But the title of the issue contains only `to the RPM packages`.

> RPM packages are used on other distributions

All of them use RPM.org-version. And here you have introduced systemd dependency, am I right?
So, the question is whether the last versions of major RPM distributions are carrying `Rpm >= 4.19` 🤷‍♂️

--- 

I also don't know about nfpm, not requesting changes and not willing to add this functionality.

> I just wanted to start with a basic example

Sure. Just skip my comments :)

**Comment by johanneskastl** [comment]
> > But the DEB
> 
> But the title of the issue contains only `to the RPM packages`.

Good catch, fixed that just now, as nfpm is used to build both DEB and RPM packages.

**Comment by 6543** [comment]
could we add example env files with default values so people can change them ...

**Comment by johanneskastl** [comment]
> could we add example env files with default values so people can change them ...

As far as I understood it there are no default values, as it highly depends on which Forge you are connecting to.

Adding a general example file that shows the syntax would be fine, but as this is all in the documentation already it does not make that much sense to duplicate that information...

**Comment by 6543** [comment]
I would still prefere it the config files would be created ... not everybody reads systemd service configs but know how to ls /etc/[service]/

Its also fine to leafe it empty or ad only a comment with a link to the docs

**Comment by johanneskastl** [comment]
> I would still prefere it the config files would be created ... not everybody reads systemd service configs but know how to ls /etc/[service]/

I have added two example files (that are named *.example) with instructions.

The units will not start, unless the file with the proper name is present. This makes it clearer / easier to find out that you need to create that file with **your** settings.

What do you think?

**Comment by 6543** [comment]
please just merge target branch into this and commit new things directly onto feature branches ...

a forcepush make it harder to review the diff from last  time i reviewed ...

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
thanks anyway :)

**Comment by kastl-ars** [comment]
> please just merge target branch into this and commit new things directly onto feature branches ...
> 
> a forcepush make it harder to review the diff from last time i reviewed ...

Sorry, rebasing is what most other projects want so I am used to it...

Thanks for taking care and merging.

**Comment by ibpan-biblioteka** [comment]
I'm confused, was this removed again shortly after being added? I don't see these files in the `main` anymore or in the .debs...

**Comment by ibpan-biblioteka** [comment]
Nvm I looked at the wrong repo's main and it seems that simply a new release wasn't made since it was merged, forget it then

**Pull request merged**